packages/core/src/render3/instructions/shared.ts

```diff
@@ -1269,7 +1269,7 @@ function saveNameToExportMap(
         exportsMap[def.exportAs[i]] = index;
       }
     }
-    if ((def as ComponentDef<any>).template) exportsMap[''] = index;
+    exportsMap[''] = index;
```

Comment:
Fix for
```
packages/core/src/render3/instructions/shared.ts(1272,9): error TS2774: This condition will always return true since the function is always defined. Did 
you mean to call it instead?
```

Listed under breaking changes in https://devblogs.microsoft.com/typescript/announcing-typescript-3-7-beta/.

Reply:
The `if` guard is there to differentiate between `DirectiveDef` and `ComponentDef`, by testing to see if `def` as a template property. This is probably still relevant, so a better fix might be

```suggestion
    if ((def as ComponentDef<any>).template !== undefined) exportsMap[''] = index;
```
or
```suggestion
    if (!!(def as ComponentDef<any>).template) exportsMap[''] = index;
```
(which is the alternative workaround in the notes)

Reply:
The cast to `ComponentDef<any>` looks odd to me since `ComponentDef` always has a template. It's means "say that def is a ComponentDef, is the template property defined?" to which it makes sense that TS would complain.

Shouldn't the cast be something different then?

Reply:
Yeah, that pattern is a bit misleading. In this case, the cast is there as an assumption of it indeed being a `ComponentDef`, then checking if `template` actually exists to confirm.

This pattern is often used in functions with type predicates:

```typescript
export function isComponentDef<T>(def: DirectiveDef<T>|ComponentDef<T>): def is ComponentDef<T> {
  return (def as ComponentDef<T>).template !== undefined;
}
```

In this case however, the compiler requires the return type to be boolean, so the original guard as written before you changed it would not have compiled in the type guard. I guess the extra function is omitted for code size (although I'm not sure on that one).

The benefit of the "assumed cast" is that there's proper reference tracking, i.e. IDEs know that the `template` property read does in fact correspond with `ComponentDef.template`, so e.g. renaming `template` will correctly update the `isComponentDef` function.